File: architect_zenghui/src/main/java/com/east/architect_zenghui/architect_35_okhttp_rxjava_retrofit/simple1/RetrofitClient.java
package com.east.architect_zenghui.architect_35_okhttp_rxjava_retrofit.simple1;

import android.util.Log;

import com.east.architect_zenghui.architect_35_okhttp_rxjava_retrofit.retrofit.ErrorHandle;
import com.east.architect_zenghui.architect_35_okhttp_rxjava_retrofit.simple2.Result;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.ObservableSource;
import io.reactivex.ObservableTransformer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by hcDarren on 2017/12/16.
 */

public class RetrofitClient {
    private final static ServiceApi mServiceApi;

    static {
        OkHttpClient okHttpClient = new OkHttpClient
                .Builder()
                .addInterceptor(new HttpLoggingInterceptor(new HttpLoggingInterceptor.Logger() {
                    @Override
                    public void log(String message) {
                        Log.e("TAG",message);
                    }
                }).setLevel(HttpLoggingInterceptor.Level.BODY))
                .build();
        // 各种套路和招式 ，发现问题解决问题，基础，源码的理解
        // 1. 没打印？
        // 2. 数据格式不一致？成功 data 是个对象，不成功 data 是个 String
        // 3. 还有就是 baseUrl 问题？ (Retrofit 找不到任何入口可以修改)
        //        3.1 不同的 baseUrl 构建不同的 Retrofit 对象 （直不应该首选）
        //        3.2 自己想办法，取巧也行走漏洞
        Retrofit retrofit = new Retrofit.Builder()
                // 访问后台接口的主路径
                .baseUrl("http://ppw.zmzxd.cn/index.php/api/v1/")
                // 添加解析转换工厂,Gson 解析，Xml解析，等等
                .addConverterFactory(GsonConverterFactory.create())
                // 添加 OkHttpClient,不添加默认就是 光杆 OkHttpClient
                .client(okHttpClient)
                // 支持RxJava Call -> Obsevrable 怎么做到的？ 1 2  采用了 Adapter 设计模式
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();

        // 创建一个 实例对象
        mServiceApi = retrofit.create(ServiceApi.class);
    }

    public static ServiceApi getServiceApi() {
        return mServiceApi;
    }

    public static <T> ObservableTransformer<Result<T>, T> transformer() {
        return new ObservableTransformer<Result<T>, T>() {
            @Override
            public ObservableSource<T> apply(Observable<Result<T>> resultObservable) {
                return resultObservable.flatMap(new Function<Result<T>, ObservableSource<T>>() {
                    @Override
                    public ObservableSource<T> apply(Result<T> tResult) throws Exception {
                        //解析不同的情况返回
                        if(tResult.isOk()){
                            //返回成功
                            return createObservable(tResult.data);
                        }else{
                            //返回失败
                            return Observable.error(new ErrorHandle.ServiceError("",tResult.getMsg()));
                        }
                    }
                }).subscribeOn(Schedulers.io()).unsubscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
            }
        };
    }

    private static <T> Observable<T> createObservable(final T data) {
        return Observable.create(new ObservableOnSubscribe<T>() {
            @Override
            public void subscribe(ObservableEmitter<T> emitter) throws Exception {
                emitter.onNext(data);
                emitter.onComplete();
            }
        });
    }
}

File: architect_zenghui/src/main/java/com/east/architect_zenghui/architect_33_retrofit1/simple2/RetrofitClient.kt
package com.east.architect_zenghui.architect_33_retrofit1.simple2

import android.util.Log
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

/**
 * |---------------------------------------------------------------------------------------------------------------|
 *  @description: Retrofit的工具类
 *  @author: jamin
 *  @date: 2020/3/16
 * |---------------------------------------------------------------------------------------------------------------|
 */
object RetrofitClient {

    lateinit var mServiceApi : ServiceApi
    init {
        var okhttpClient = OkHttpClient.Builder()
            .addInterceptor(HttpLoggingInterceptor(object : HttpLoggingInterceptor.Logger {
                override fun log(message: String) {
                    Log.e("TAG", message)
                }
            }).apply {
                this.level = HttpLoggingInterceptor.Level.BODY
            })
            .build()
        // 各种套路和招式 ，发现问题解决问题，基础，源码的理解
        // 1. 没打印？
        // 2. 数据格式不一致？成功 data 是个对象，不成功 data 是个 String
        // 3. 还有就是 baseUrl 问题？ (Retrofit 找不到任何入口可以修改)
        //        3.1 不同的 baseUrl 构建不同的 Retrofit 对象 （直不应该首选）
        //        3.2 自己想办法，取巧也行走漏洞

        var retrofit = Retrofit.Builder()
            // 访问后台接口的主路径
            .baseUrl("http://192.168.6.143:8080/OkHttpServer/")
            // 添加解析转换工厂,Gson 解析，Xml解析，等等
            .addConverterFactory(GsonConverterFactory.create())
            // 添加 OkHttpClient,不添加默认就是 光杆 OkHttpClient
            .client(okhttpClient)
            .build()

        //创建一个实例对象
        mServiceApi = retrofit.create(ServiceApi::class.java)
    }

    
    fun getServiceApi(): ServiceApi = mServiceApi
}